File: golang/components_test.go
package httpsig

import (
	"net/http"
	"net/url"
	"testing"
)

func testRequestMessage() Message {
	u, _ := url.Parse("https://example.com/foo?param=Value&Pet=dog")
	return &RawMessage{
		IsReq:     true,
		Meth:      "POST",
		TargetURI: u,
		Headers: http.Header{
			"Host":           {"example.com"},
			"Date":           {"Tue, 20 Apr 2021 02:07:55 GMT"},
			"Content-Type":   {"application/json"},
			"Content-Digest": {"sha-512=:WZDPaVn/7XgHaAy8pmojAkGWoRx2UFChF41A2svX+TaPm+AbwAgBWnrIiYllu7BNNyealdVLvRwEmTHWXvJwew==:"},
			"Content-Length": {"18"},
		},
	}
}

func TestExtractDerivedComponents(t *testing.T) {
	msg := testRequestMessage()

	tests := []struct {
		component ComponentIdentifier
		want      string
	}{
		{Component("@method"), "POST"},
		{Component("@path"), "/foo"},
		{Component("@query"), "?param=Value&Pet=dog"},
		{Component("@authority"), "example.com"},
		{Component("@scheme"), "https"},
		{Component("@target-uri"), "https://example.com/foo?param=Value&Pet=dog"},
		{Component("@request-target"), "/foo?param=Value&Pet=dog"},
		{QueryParam("Pet"), "dog"},
		{QueryParam("param"), "Value"},
	}

	for _, tt := range tests {
		t.Run(tt.component.Name, func(t *testing.T) {
			got, err := extractComponent(tt.component, msg, nil)
			if err != nil {
				t.Fatalf("extractComponent: %v", err)
			}
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestExtractHeaders(t *testing.T) {
	msg := testRequestMessage()

	tests := []struct {
		name string
		want string
	}{
		{"content-type", "application/json"},
		{"date", "Tue, 20 Apr 2021 02:07:55 GMT"},
		{"content-length", "18"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := extractComponent(Component(tt.name), msg, nil)
			if err != nil {
				t.Fatalf("extractComponent: %v", err)
			}
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestExtractMultiValueHeader(t *testing.T) {
	u, _ := url.Parse("https://example.org/demo")
	msg := &RawMessage{
		IsReq:     true,
		Meth:      "GET",
		TargetURI: u,
		Headers: http.Header{
			"Accept": {"application/json", "*/*"},
		},
	}

	got, err := extractComponent(Component("accept"), msg, nil)
	if err != nil {
		t.Fatalf("extractComponent: %v", err)
	}
	want := "application/json, */*"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestExtractMissingHeader(t *testing.T) {
	msg := testRequestMessage()
	_, err := extractComponent(Component("x-not-present"), msg, nil)
	if err == nil {
		t.Error("expected error for missing header")
	}
}

func TestExtractStatusOnRequest(t *testing.T) {
	msg := testRequestMessage()
	_, err := extractComponent(Component("@status"), msg, nil)
	if err == nil {
		t.Error("expected error for @status on request")
	}
}

func TestExtractMethodOnResponse(t *testing.T) {
	msg := &RawMessage{
		IsReq:  false,
		Status: 200,
		Headers: http.Header{
			"Content-Type": {"application/json"},
		},
	}
	_, err := extractComponent(Component("@method"), msg, nil)
	if err == nil {
		t.Error("expected error for @method on response")
	}
}

func TestExtractStatus(t *testing.T) {
	msg := &RawMessage{
		IsReq:  false,
		Status: 200,
		Headers: http.Header{
			"Content-Type": {"application/json"},
		},
	}
	got, err := extractComponent(Component("@status"), msg, nil)
	if err != nil {
		t.Fatalf("extractComponent: %v", err)
	}
	if got != "200" {
		t.Errorf("got %q, want %q", got, "200")
	}
}

func TestExtractQueryParamMissing(t *testing.T) {
	msg := testRequestMessage()
	_, err := extractComponent(QueryParam("missing"), msg, nil)
	if err == nil {
		t.Error("expected error for missing query param")
	}
}

File: typescript/test/components.test.ts
import { describe, it, expect } from "vitest";
import {
  component,
  queryParam,
  componentReq,
  extractComponent,
} from "../src/index.js";
import type { HttpMessage } from "../src/index.js";

function makeRequest(
  method: string,
  url: string,
  headers: Record<string, string | string[]> = {},
): HttpMessage {
  const headerMap = new Map<string, string[]>();
  for (const [name, value] of Object.entries(headers)) {
    const key = name.toLowerCase();
    headerMap.set(key, Array.isArray(value) ? value : [value]);
  }
  return {
    isRequest: true,
    method,
    url: new URL(url),
    headerValues: (name: string) => headerMap.get(name.toLowerCase()) ?? [],
  };
}

describe("derived components", () => {
  const msg = makeRequest(
    "POST",
    "https://example.com/foo?param=Value&Pet=dog",
    { "content-type": "application/json" },
  );

  it("@method", () => {
    expect(extractComponent(component("@method"), msg)).toBe("POST");
  });

  it("@path", () => {
    expect(extractComponent(component("@path"), msg)).toBe("/foo");
  });

  it("@authority", () => {
    expect(extractComponent(component("@authority"), msg)).toBe("example.com");
  });

  it("@scheme", () => {
    expect(extractComponent(component("@scheme"), msg)).toBe("https");
  });

  it("@target-uri", () => {
    expect(extractComponent(component("@target-uri"), msg)).toBe(
      "https://example.com/foo?param=Value&Pet=dog",
    );
  });

  it("@request-target", () => {
    expect(extractComponent(component("@request-target"), msg)).toBe(
      "/foo?param=Value&Pet=dog",
    );
  });

  it("@query", () => {
    expect(extractComponent(component("@query"), msg)).toBe(
      "?param=Value&Pet=dog",
    );
  });

  it("@query-param", () => {
    expect(extractComponent(queryParam("Pet"), msg)).toBe("dog");
  });
});

describe("header components", () => {
  it("extracts single header", () => {
    const msg = makeRequest("GET", "https://example.com", {
      "content-type": "application/json",
    });
    expect(extractComponent(component("content-type"), msg)).toBe(
      "application/json",
    );
  });

  it("combines multi-value headers", () => {
    const msg = makeRequest("GET", "https://example.com", {
      accept: ["application/json", "*/*"],
    });
    expect(extractComponent(component("accept"), msg)).toBe(
      "application/json, */*",
    );
  });

  it("throws for missing header", () => {
    const msg = makeRequest("GET", "https://example.com");
    expect(() => extractComponent(component("x-missing"), msg)).toThrow(
      "missing component",
    );
  });
});

describe("response components", () => {
  it("@status", () => {
    const msg: HttpMessage = {
      isRequest: false,
      statusCode: 200,
      headerValues: () => [],
    };
    expect(extractComponent(component("@status"), msg)).toBe("200");
  });

  it("@method on response throws", () => {
    const msg: HttpMessage = {
      isRequest: false,
      statusCode: 200,
      headerValues: () => [],
    };
    expect(() => extractComponent(component("@method"), msg)).toThrow(
      "@method on response",
    );
  });
});
